### ISSUE TITLE
Incorrect Transformation

### ISSUE THREAD

**Issue by SukkaW** [lxsmnsyc/forgetti#24]
**Forgetti Version**

0.5.0

**Input**

```jsx
'use client';

import { useState, useCallback, memo } from 'react';
import { SMALLEST_GIF } from '@/lib/shared/constant';
import { useIntersection } from '@/hooks/use-intersection';

const Thumbnail = ({
  src,
  lazy
}) => {
  const [setIntersection, isIntersected, resetIsIntersected] = useIntersection({
    rootMargin: '200px'
  });

  const [previousSrc, setPreviousSrc] = useState<string | null>(src);

  let srcOrWebpHasChangedSinceLastRender = false;
  if (previousSrc !== src) {
    setPreviousSrc(src);
    srcOrWebpHasChangedSinceLastRender = true;
  }
  if (srcOrWebpHasChangedSinceLastRender) {
    resetIsIntersected();
  }

  const isVisible = isIntersected && !srcOrWebpHasChangedSinceLastRender;
  console.log(isIntersected, srcOrWebpHasChangedSinceLastRender, isVisible);

  return (
    <img
      ref={useCallback((el) => {
        if (lazy) {
          setIntersection(el);
        }
      }, [lazy, setIntersection])}
      src={isVisible ? src : SMALLEST_GIF}
    />
  );
};

export default memo(Thumbnail);
```

**Output**

```jsx
'use client';
import { jsx as _jsx } from "react/jsx-runtime";
import { useMemo as _useMemo } from "react";
import { $$cache as _$$cache } from "forgetti/runtime";
import { $$branch as _$$branch } from "forgetti/runtime";
import { $$equals as _$$equals } from "forgetti/runtime";
import { useState, memo } from 'react';
import { SMALLEST_GIF } from '@/lib/shared/constant';
import { useIntersection } from '@/hooks/use-intersection';
const Thumbnail = ({ src , lazy  })=>{
    let _c = _$$cache(_useMemo, 21), _hoisted = useIntersection(0 in _c ? _c[0] : _c[0] = {
        rootMargin: '200px'
    });
    const [setIntersection, isIntersected, resetIsIntersected] = _$$equals(_c, 1, _hoisted) ? _c[1] : _c[1] = _hoisted;
    let _eq2 = _$$equals(_c, 2, src), _v3 = _eq2 ? _c[2] : _c[2] = src, _hoisted2 = useState(_v3);
    const [previousSrc, setPreviousSrc] = _$$equals(_c, 3, _hoisted2) ? _c[3] : _c[3] = _hoisted2;
    let srcOrWebpHasChangedSinceLastRender = false, _eq4 = _$$equals(_c, 4, previousSrc), _v5 = _eq4 ? _c[4] : _c[4] = previousSrc;
    if (_eq4 && _eq2 ? _c[5] : _c[5] = _v5 !== _v3) {
        let _c2 = _$$branch(_c, 6, 2);
        _eq2 ? _c2[0] : _c2[0] = setPreviousSrc(_v3);
        let _hoisted3 = srcOrWebpHasChangedSinceLastRender = true;
        _$$equals(_c2, 1, _hoisted3) ? _c2[1] : _c2[1] = _hoisted3;
    }
    let _eq7 = _$$equals(_c, 7, srcOrWebpHasChangedSinceLastRender), _v9 = _eq7 ? _c[7] : _c[7] = srcOrWebpHasChangedSinceLastRender;
    if (_v9) {
        let _c3 = _$$branch(_c, 8, 1);
        0 in _c3 ? _c3[0] : _c3[0] = resetIsIntersected();
    }
    let _eq8 = _$$equals(_c, 9, isIntersected), _v12 = _eq8 ? _c[9] : _c[9] = isIntersected, _v11;
    if (_v12) {
        let _c4 = _$$branch(_c, 10, 1);
        _v11 = _eq7 ? _c4[0] : _c4[0] = !_v9;
    } else _v11 = _v12;
    const isVisible = _v11;
    let _eq9 = _$$equals(_c, 11, isVisible), _v14 = _eq9 ? _c[11] : _c[11] = isVisible;
    _eq8 && _eq7 && _eq9 ? _c[12] : _c[12] = console.log(_v12, _v9, _v14);
    let _eq11 = _$$equals(_c, 13, lazy), _v16 = _eq11 ? _c[13] : _c[13] = lazy, _eq12 = _$$equals(_c, 14, setIntersection), _v17 = _eq12 ? _c[14] : _c[14] = setIntersection, _eq13 = _eq11 && _eq12, _v18 = _eq13 ? _c[15] : _c[15] = [
        _v16,
        _v17
    ], _v19 = _eq13 ? _c[16] : _c[16] = (el)=>{
        if (lazy) {
            setIntersection(el);
        }
    }, _eq14 = _$$equals(_c, 17, _v19), _v20 = _eq14 ? _c[17] : _c[17] = _v19, _v21;
    if (_v14) {
        _v21 = _v3;
    } else {
        let _c5 = _$$branch(_c, 18, 1);
        _v21 = 0 in _c5 ? _c5[0] : _c5[0] = SMALLEST_GIF;
    }
    let _eq15 = _$$equals(_c, 19, _v21), _v23 = _eq15 ? _c[19] : _c[19] = _v21;
    return _eq14 && _eq15 ? _c[20] : _c[20] = /*#__PURE__*/ _jsx("img", {
        ref: _v20,
        src: _v23
    });
};
export default /*#__PURE__*/ memo(Thumbnail);
```

----

I was writing a thumbnail component with lazyload support. I have added the `console.log(isIntersected, srcOrWebpHasChangedSinceLastRender, isVisible);` to debug.

Expected logging `false false false` then `true false true`, but get `false false false` then `true false undefined`.

**Comment by SukkaW** [comment]
Found out why. Notice the line `let _c4 = _$$branch(_c, 10, 1);`. When `_eq7` becomes true, value from `_c4[0]` is used, but `_c4` (which is `_c[10]`) is actually `[empty]`:

<img width="432" alt="image" src="https://github.com/lxsmnsyc/forgetti/assets/40715044/8ddfb11e-7161-4968-870b-88cdc242085f">

**Comment by lxsmnsyc** [comment]
hmmm interesting. I guess it's because of how the `&&` was short-circuited on first render

**Comment by lxsmnsyc** [comment]
Should be fixed in 0.5.2

**Comment by SukkaW** [comment]
> Should be fixed in 0.5.2

Just installed 0.5.3 and confirmed the issue is fixed, thanks!
